### PR TITLE
fix(http1): poll_loop writes when ready

### PR DESCRIPTION
**Pull request by lthiery**
I ran into some lockups running hyper with some custom futures. If one of my futures is ready when polled, the waker is never signaled and I think this uncovered a logical issue with the http1 poll_loop. That it to say, **if the poll_write is demonstrating readiness to write and the connection has readiness to write, we should do that or else hyper may stall**.

I built a pathological example "ready_stream.rs" - you can run it from this branch with: `RUSTFLAGS='--cfg hyper_unstable_tracing' cargo test --test ready_stream --features full,tracing`.

I provided a proposed patch for the poll_loop, but I'm open to other angles on this.

### CODE REVIEW

File: tests/ready_stream.rs [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+use std::convert::Infallible;
+use std::io;
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+use tokio::sync::mpsc;
+use tracing::{error, info};
+
+pin_project! {
+    #[derive(Debug)]
+    pub struct TxReadyStream {
```
[the remark this answers is not in the document]
Comment (lthiery) [reply tests/ready_stream.rs]: It's a little difficult to simplify because I needed to create a future that would deterministically trigger this edge case.

What's "weird" about this future mock is that it can return Poll::Pending once, but then reach completion on a future poll without the waker being woken, so I don't think it ever happens in Tokio for example but it does happen (probabilistically) with my actual futures that this is modeling. And as far as I know it's no violation of the Futures contract to do so.

Anyway, I'm not sure I can create the situation with less or more simple code. That being said, maybe a fixture like this is generalizable for the project? Creating a "connection" that's purely channel based software could allow other edge cases to be explored. Partial reads and writes or different sequence of readiness.

All that being said, I'll still take another critical pass at what I've done here next week and I'll see what I can do to simplify.

**Comment by lthiery** [comment]
It's an interesting loop and I think with this edge case handled it works well. It would be nice to expose the 16 as a config (as noted in the comments) and maybe I'll PR that later.

One thing that I considered while looking at it is that maybe doing some tricks with custom wakers could be nice? Similar to FuturesUnordered, the waker could indicate who is actually ready for polling. The juice is probably not worth the squeeze (or perhaps even an efficiency loss with all the bookkeeping?) as I know FuturesUnordered is intended for some level of scale and three is probably not it.

**Comment by seanmonstar** [comment]
Yea, hm, if it's a complicated edge case, perhaps that's fine as-is, then.
